packages/preferences/src/browser/preference-transaction-manager.ts

```diff
@@ -0,0 +1,203 @@
+/********************************************************************************
+ * Copyright (C) 2021 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { CancellationError, Emitter, Event, MaybePromise, MessageService, nls, WaitUntilEvent } from '@theia/core';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { inject, injectable, postConstruct } from '@theia/core/shared/inversify';
+import { PreferenceScope } from '@theia/core/lib/common/preferences/preference-scope';
+import URI from '@theia/core/lib/common/uri';
+import { MonacoEditorModel } from '@theia/monaco/lib/browser/monaco-editor-model';
+import { Mutex, MutexInterface } from 'async-mutex';
+import { MonacoTextModelService } from '@theia/monaco/lib/browser/monaco-text-model-service';
+import { MonacoJSONCEditor } from './monaco-jsonc-editor';
+import { EditorManager } from '@theia/editor/lib/browser/editor-manager';
+
+export interface OnWillConcludeEvent<T> extends WaitUntilEvent {
+    status: T | false;
+}
+
+@injectable()
+export abstract class Transaction<Arguments extends unknown[], Result = unknown, Status = unknown> {
+    protected _open = true;
+    get open(): boolean {
+        return this._open;
+    }
+    protected _result = new Deferred<Result | false>();
+    get result(): Promise<Result | false> {
+        return this._result.promise;
+    }
+    protected readonly queue = new Mutex(new CancellationError);
```

Comment:
Nitpick: Missing parenthesis. We should probably activate the ESLint rule for that, I've seen that in some other PR reviews as well.

Reply:
The omission was deliberate but completely arbitrary. I've added the parentheses. 👍